### PR TITLE
feat(aws/cloudformation): Add cloudformation caching agent and provider

### PR DESCRIPTION
**Pull request by xavileon**
This patch adds two components necessary to specify and implement a
cloud formation stage in a pipeline.

* Cloudformation caching agent: it's responsible to periodically query
aws to cache the relevant information from a cloudformation stack,
including the status whenever it's in an errored state.

* Cloudformation provider: it provides an API to query the cache and get
a cloudformation stack given a stackId or list existing stacks for an
aws account and (optionally) a region.

The cloudformation controller that will be used by Orca as well as the
task that deploys this cloud formation will be added in separate PRs.

This is the first PR of a series to address https://github.com/spinnaker/spinnaker/issues/736

Signed-off-by: Xavi León <xavi.leon@schibsted.com>

### CODE REVIEW

**Comment by xavileon** [comment]
Regarding this comment https://github.com/spinnaker/clouddriver/pull/3278#issuecomment-454773226 All the code in the `clouddriver-aws`module is in groovy and added the new code in groovy for consistency, is it preferable to move it to Java in any case?

**Comment by ethanfrogers** [comment]
@xavileon yes, any new code should be written in something non-Groovy, however if there is already another language being used in the module (Kotlin, Java, etc) you should use that. I believe we try to keep modules to no more than 2 languages.

**Comment by xavileon** [comment]
> @xavileon yes, any new code should be written in something non-Groovy, however if there is already another language being used in the module (Kotlin, Java, etc) you should use that. I believe we try to keep modules to no more than 2 languages.

ok, thanks! I'll move the code to Java then 👍

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProvider.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    String filter = Keys.getCloudFormationKey(stackId, "*", "*");
+    log.debug(String.format("Get stack with filter %s", filter));
+    return loadResults(cacheView.filterIdentifiers(CLOUDFORMATION.getNs(), filter)).stream().findFirst().get();
+  }
+
+  List<AmazonCloudFormation> loadResults(Collection<String> identifiers) {
+    return cacheView.getAll(CLOUDFORMATION.getNs(), identifiers, RelationshipCacheFilter.none())
+      .stream()
+      .map(data -> {
+        log.debug(String.format("Cloud formation cached properties %s", data.getAttributes()));
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProvider.java]: You don't need to use `String.format` in log calls...

```java
log.debug("Cloud formation cached properties {}", data.getAttributes());
```

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgentSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    cache = agent.loadData(providerCache)
+    results = cache.cacheResults[Keys.Namespace.CLOUDFORMATION.ns]
+
+    then:
+    1 * acp.getAmazonCloudFormation(_, _) >> amazonCloudFormation
+    1 * amazonCloudFormation.describeStacks() >> stackResults
+    1 * stackResults.stacks >> [ stack1, stack2 ]
+
+    results.find { it.id == Keys.getCloudFormationKey("stack1", "region", "account") }
+    results.find { it.id == Keys.getCloudFormationKey("stack2", "region", "account") }
```
Comment (robzienert) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgentSpec.groovy]: Are you wanting to actually assert on the presence of these? I don't think these two lines don't do anything currently.

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgentSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  void "should include stack status reason when state is ROLLBACK_COMPLETE (failed)"() {
+    given:
+    def amazonCloudFormation = Mock(AmazonCloudFormation)
+    def stack = new Stack().withStackId("stack1").withStackStatus("ROLLBACK_COMPLETE")
+    def stackResults = Mock(DescribeStacksResult)
+    def stackEvent = new StackEvent().withResourceStatus("CREATE_FAILED").withResourceStatusReason("who knows")
+    def stackEventResults = Mock(DescribeStackEventsResult)
+    def cache
+    def results
```
Comment (robzienert) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgentSpec.groovy]: I'd just put the `def` of `cache` and `results` in the `when` stanza.

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    then:
+    result == cloudFormations.findAll { it.accountId == accountId } as Set
+
+    where:
+    accountId  || count
+    "account1" || 2
+    "account2" || 1
+    "unknown"  || 0
+    null       || 0
```
Comment (robzienert) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy]: Are you wanting to assert anything with `count`?

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudFormation.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.clouddriver.model;
+
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A representation of a CloudDriver stack
```
Comment (robzienert) [inline clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudFormation.java]: `s/Driver/Formation/`

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProvider.java (hunk 1 above)
Comment (xavileon) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProvider.java]: done in https://github.com/spinnaker/clouddriver/pull/3278/commits/85c6eadada2947091e4559d29078e29a08a40ba6

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy (hunk 4 above)
Comment (xavileon) [reply clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy]: good question, it's a left over from a previous way of testing. I've changed the test a bit to make it more clear and explicit on the assertions.
done in https://github.com/spinnaker/clouddriver/pull/3278/commits/7455e0a53b9a12e6eed371e5cb7b9f4a895a7ec2

**Comment by xavileon** [comment]
@robzienert thanks for the review 👍 addressed the comments about the logs in https://github.com/spinnaker/clouddriver/pull/3278/commits/85c6eadada2947091e4559d29078e29a08a40ba6 and the tests in https://github.com/spinnaker/clouddriver/pull/3278/commits/7455e0a53b9a12e6eed371e5cb7b9f4a895a7ec2

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgent.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.clouddriver.aws.provider.agent;
+
+import com.amazonaws.services.cloudformation.AmazonCloudFormation;
+import com.amazonaws.services.cloudformation.model.*;
+import com.amazonaws.services.cloudformation.model.Stack;
+import com.netflix.spinnaker.cats.agent.*;
```
Comment (jervi) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonCloudFormationCachingAgent.java]: Split into separate imports

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonCloudFormationSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    cf.stackId == "stackId"
+    cf.tags == [tag1: "tag1", tag2: "tag2"]
+    cf.outputs == [out1: "out1", out2: "out2"]
+    cf.stackName == "stackName"
+    cf.region == "region"
+    cf.accountName == "accountName"
+    cf.accountId == "accountId"
+    cf.stackStatus == "stackStatus"
+    cf.stackStatusReason == "stackStatusReason"
+  }
```
Comment (jervi) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/model/AmazonCloudFormationSpec.groovy]: You can use the `with` method in spock:
```
    with(cf) {
      stackId == "stackId"
      tags == [tag1: "tag1", tag2: "tag2"]
      outputs == [out1: "out1", out2: "out2"]
      stackName == "stackName"
      region == "region"
      accountName == "accountName"
      accountId == "accountId"
      stackStatus == "stackStatus"
      stackStatusReason == "stackStatusReason"
    }
```

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    where:
+    accountId  || expected
+    "account1" || [ stack1, stack2 ] as Set
+    "account2" || [ stack3 ] as Set
+    "unknown"  || [] as Set
+    null       || [] as Set
+  }
+
+  @Unroll
+  void "list all cloud formations by account and region"() {
```
Comment (jervi) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/provider/view/AmazonCloudFormationProviderSpec.groovy]: Tip: You can add variables to the name of unrolled tests:
`void "list all cloud formations by account '#account' and region '#region'"() {`

**Pull request merged**